x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.ts

```diff
@@ -205,6 +211,60 @@ export const generateId = (
   ruleId: string
 ): string => createHash('sha256').update(docIndex.concat(docId, version, ruleId)).digest('hex');
 
+// TODO: do we need to include version in the id? If it does matter then we should include it in signal.parents as well
```

Comment:
In the other parts of the code we were using `_version` for deduplication in case the user changes a document's contents (which really shouldn't happen with beats since it is all time based documents) and we wanted to create a new signal off of that change so it is not missed.

However, using `_version` might be a mistake and left over from a time ago.

Lately we saw from here:
https://www.elastic.co/guide/en/elasticsearch/reference/current/optimistic-concurrency-control.html
https://discuss.elastic.co/t/relation-between-version-seq-no-and-primary-term/179647/3
https://github.com/elastic/elasticsearch/issues/19269#issuecomment-488598561

To uniquely identify the document it might be better to use  `_seq_no and _primary_term` instead.

Reply:
You may be referencing something else, but `_version` is used for saved objects, and it encapsulates both `seq_no` and `primary_term`. For data indexes, you need both to identify a unique doc for OOC, as Frank pointed out. If there's some other `_version`, then I'm not familiar with it.

Reply:
My concern with using `_version` is that updates to documents don't necessarily mean we should generate a new signal. Particularly in the signals-on-signals case where a signal might be updated with status or user assignment information we definitely don't want to create another new signal-on-signal just because the original one was updated. Right now we don't actually use `_version` afaik when updating signal status but we probably should use it in the future. 

I'd be more comfortable assuming that we only care about running detection engine logic against each document once and require that any substantive changes to a document (which would also increment the `_version`) must be made as a new document instead in order for the detection engine to detect them. This approach also fits in with the time range filters we apply - since we're only looking at new documents, substantive changes that are inserted as new documents will be queried by the DE, whereas if we support modifying documents in place then the modified documents may or may not be queried depending on the time range and when the next rule execution runs.